integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ExpressionDependencyCollector.java

```diff
@@ -52,52 +70,176 @@ static void collectFromNode(ColumnLevelLineageContext context, LogicalPlan node)
       expressions.addAll(
           ScalaConversionUtils.<NamedExpression>fromSeq(((Aggregate) node).aggregateExpressions()));
+    } else if (node instanceof Join) {
+      Option<Expression> condition = ((Join) node).condition();
+      if (condition.isDefined()) {
+        datasetTransformation = Optional.of(TransformationInfo.indirect(JOIN));
+        datasetDependencies.add(condition.get());
+      }
+    } else if (node instanceof Filter) {
+      datasetDependencies.add(((Filter) node).condition());
+      datasetTransformation = Optional.of(TransformationInfo.indirect(FILTER));
+    } else if (node instanceof Sort) {
+      datasetDependencies.addAll(ScalaConversionUtils.<SortOrder>fromSeq(((Sort) node).order()));
+      datasetTransformation = Optional.of(TransformationInfo.indirect(SORT));
     } else if (node instanceof LogicalRelation) {
       if (((LogicalRelation) node).relation() instanceof JDBCRelation) {
         JdbcColumnLineageCollector.extractExpressionsFromJDBC(context, node);
       }
     }
-
     expressions.stream()
         .forEach(
-            expr -> traverseExpression((Expression) expr, expr.exprId(), context.getBuilder()));
+            expr ->
+                traverseExpression(
+                    (Expression) expr,
+                    expr.exprId(),
+                    TransformationInfo.identity(),
+                    context.getBuilder()));
+
+    datasetTransformation.ifPresent(
+        dt -> {
+          ExprId exprId = NamedExpression.newExprId();
+          context.getBuilder().addDatasetDependency(exprId);
+          datasetDependencies.forEach(e -> traverseExpression(e, exprId, dt, context.getBuilder()));
+        });
   }
 
   public static void traverseExpression(
       Expression expr, ExprId outputExprId, ColumnLevelLineageBuilder builder) {
-    if (expr instanceof NamedExpression
-        && !((NamedExpression) expr).exprId().equals(outputExprId)) {
-      builder.addDependency(outputExprId, ((NamedExpression) expr).exprId());
-    }
+    traverseExpression(expr, outputExprId, TransformationInfo.identity(), builder);
+  }
+
+  public static void traverseExpression(
+      Expression expr,
+      ExprId outputExprId,
+      TransformationInfo transformationInfo,
+      ColumnLevelLineageBuilder builder) {
 
-    // discover children expression -> handles UnaryExpressions like Alias
-    if (expr.children() != null) {
-      ScalaConversionUtils.<Expression>fromSeq(expr.children()).stream()
-          .forEach(child -> traverseExpression(child, outputExprId, builder));
+    if (expr instanceof AttributeReference) {
+      AttributeReference attRef = (AttributeReference) expr;
+      if (!attRef.exprId().equals(outputExprId)) {
+        builder.addDependency(outputExprId, attRef.exprId(), transformationInfo);
+      }
+    } else if (expr instanceof Alias) {
+      handleAlias((Alias) expr, outputExprId, transformationInfo, builder);
```

Comment:
couldn't you have a single `handle` method and let java chose implementation based on the given type?

Reply:
oh crap, you're right! Nice!